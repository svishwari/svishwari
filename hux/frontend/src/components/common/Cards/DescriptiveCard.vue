<template>
  <v-card
    :outlined="disabled"
    class="descriptive-card align-center text-center rounded-lg card-space mb-6"
    :class="{ 'in-active': disabled }"
    :height="height"
    :width="width"
    :to="to"
  >
    <div v-if="$slots.top" class="card-status pa-3 pb-0">
      <slot name="top" />
      <v-menu close-on-click>
        <template #activator="{ on, attrs }">
          <v-icon
            v-if="actionMenu"
            class="d-flex float-right"
            v-bind="attrs"
            color="primary"
            v-on="on"
          >
            mdi-dots-vertical
          </v-icon>
        </template>
        <div class="black--text text-darken-4 cursor-pointer white">
          <slot name="action-menu-options"></slot>
        </div>
      </v-menu>
      <div v-if="comingSoon" class="coming-soon d-flex float-right mt-n4">
        Coming soon!
      </div>
    </div>

    <div v-if="icon" class="d-flex justify-center mt-2 mr-8">
      <div class="dot">
        <logo
          v-if="logoOption"
          :type="icon"
          :size="32"
          :color="iconColor"
          class="d-block"
        />
        <icon
          v-else
          :type="icon"
          :size="32"
          :color="iconColor"
          class="d-block"
        />
      </div>
    </div>

    <tooltip nudge-right="100px" min-width="auto !important">
      <template #label-content>
        <div
          class="text-h4 px-6 pb-1 pt-2 text-ellipsis d-block title text-h4"
          :class="disabled ? 'black--text text--darken-4' : 'primary--text'"
          :style="{ 'padding-top': !icon ? '56px' : null }"
          data-e2e="card-title"
        >
          {{ title }}
        </div>
      </template>
      <template #hover-content>
        <span class="black--text text--lighten-4">{{ title }}</span>
      </template>
    </tooltip>

    <tooltip nudge-right="100px" min-width="auto !important">
      <template #label-content>
        <div
          class="px-3 d-block description text-body-2 black--text"
          :style="{ 'padding-top': !icon ? '22px' : null }"
          data-e2e="card-description"
        >
          {{ description }}
        </div>
      </template>
      <template #hover-content>
        <span class="black--text text--lighten-4">{{ description }}</span>
      </template>
    </tooltip>

    <div v-if="$slots.default" class="px-3 pt-2">
      <slot />
    </div>
  </v-card>
</template>

<script>
import Icon from "@/components/common/Icon"
import Tooltip from "@/components/common/Tooltip"
import Logo from "@/components/common/Logo"

export default {
  name: "DescriptiveCard",

  components: {
    Icon,
    Tooltip,
    Logo,
  },

  props: {
    icon: {
      type: String,
      required: false,
    },
    title: {
      type: String,
      required: false,
      default: "Model Name",
    },
    description: {
      type: String,
      required: false,
      default: "Descriptive text for the model item chosen above",
    },
    disabled: {
      type: Boolean,
      required: false,
      default: false,
    },
    actionMenu: {
      type: Boolean,
      required: false,
      default: false,
    },
    comingSoon: {
      type: Boolean,
      required: false,
      default: false,
    },
    height: {
      type: [Number, String],
      required: true,
    },
    width: {
      type: [Number, String],
      required: true,
    },
    to: {
      type: Object,
      required: false,
      default: () => {},
    },
    dotOption: {
      type: String,
      required: false,
      default: "Activate",
    },
    logoOption: {
      type: Boolean,
      required: false,
      default: false,
    },
    iconColor: {
      type: String,
      required: false,
      default: "Primary",
    },
  },
}
</script>

<style lang="scss" scoped>
.descriptive-card {
  @extend .box-shadow-5;
  color: var(--v-black-darken4);
  font-weight: normal;
  transition: box-shadow 0.2s;
  cursor: pointer;

  &:hover {
    @extend .box-shadow-3;
  }
  &.in-active {
    cursor: default;
    background-color: var(--v-primary-lighten1);
    &:hover {
      @extend .box-shadow-5;
    }
    .coming-soon {
      width: 104px;
      height: 28px;
      margin-top: -12px;
      position: absolute;
      right: 0;
      background: #e2eaec;
      border-radius: 0px 12px;
      padding: 4px;
    }
  }
  .dot {
    width: 60px;
    height: 60px;
    border-radius: 50%;
    @extend .box-shadow-1;
    background: var(--v-white-base);
    text-align: -webkit-center;
    padding: 14px;
  }
  .description {
    min-height: 36px;
    -webkit-box-orient: vertical !important;
    -webkit-line-clamp: 2 !important;
    overflow: hidden !important;
    display: -webkit-box !important;
  }
}
.card-space {
  margin-right: 24px !important;
}
</style>
