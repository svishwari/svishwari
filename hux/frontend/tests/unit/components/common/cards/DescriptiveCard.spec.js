import DescriptiveCard from "@/components/common/Cards/DescriptiveCard"
import { shallowMount } from "@vue/test-utils"

describe("Cards", () => {
  describe("Descriptive Card", () => {
    test("Displays custom properties", () => {
      const customProps = {
        icon: "model-unsubscribe",
        title: "Propensity to Unsubscribe",
        type: "Models",
        description:
          "Propensity of a customer making a purchase after receiving an email.",
      }

      const wrapper = shallowMount(DescriptiveCard, {
        propsData: customProps,
      })
      // TODO: since the content is now moved inside tooltip it is stubbing
      // the tooltip and not mounting the title and description in shallowMount
      // mounting the element needs to load alot of dependencies which needs further investigation.
      expect(wrapper).toMatchSnapshot()
    })

    test("Displays custom slots", () => {
      const customProps = {
        type: "Modules",
      }
      const customSlots = {
        default: "<p>Default content</p>",
        top: "<div>Top spot</div>",
      }

      const wrapper = shallowMount(DescriptiveCard, {
        slots: customSlots,
        propsData: customProps,
      })

      expect(wrapper.find("p").exists()).toBe(true)
      expect(wrapper.find("div").exists()).toBe(true)
      expect(wrapper).toMatchSnapshot()
    })
  })
})
