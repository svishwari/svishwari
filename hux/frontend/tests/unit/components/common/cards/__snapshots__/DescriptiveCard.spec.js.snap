// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cards Descriptive Card Displays custom properties 1`] = `
<v-card-stub loaderheight="4" tag="div" height="255" width="280" class="descriptive-card align-center text-center rounded-lg card-space mb-6 non-interactable">
  <!---->
  <div class="d-flex justify-center mt-2 mr-8">
    <div class="dot" style="padding: 14px;">
      <icon-stub type="model-unsubscribe" size="32" color="Primary" fillopacity="1" strokeopacity="1" class="d-block"></icon-stub>
    </div>
  </div>
  <tooltip-stub contentclass="" maxwidth="auto" minwidth="auto !important" nudgeright="100px"></tooltip-stub>
  <tooltip-stub contentclass="" maxwidth="auto" minwidth="auto !important" nudgeright="100px"></tooltip-stub>
  <!---->
</v-card-stub>
`;

exports[`Cards Descriptive Card Displays custom slots 1`] = `
<v-card-stub loaderheight="4" tag="div" height="255" width="280" class="descriptive-card align-center text-center rounded-lg card-space mb-6 non-interactable">
  <div class="card-status pa-3 pb-0">
    <div>Top spot</div>
    <v-menu-stub opendelay="0" closedelay="0" contentclass="" maxwidth="auto" nudgebottom="0" nudgeleft="0" nudgeright="0" nudgetop="0" nudgewidth="0" openonclick="true" closeonclick="true" closeoncontentclick="true" maxheight="auto" origin="top left" transition="v-menu-transition">
      <div class="black--text text-darken-4 cursor-pointer white"></div>
    </v-menu-stub>
    <!---->
  </div>
  <!---->
  <tooltip-stub contentclass="" maxwidth="auto" minwidth="auto !important" nudgeright="100px"></tooltip-stub>
  <tooltip-stub contentclass="" maxwidth="auto" minwidth="auto !important" nudgeright="100px"></tooltip-stub>
  <div class="px-3 pt-2">
    <p>Default content</p>
  </div>
</v-card-stub>
`;
